Evaluation/eval_tools.py

```diff
@@ -217,7 +217,7 @@ def create_roc_curve(self, df):
         return roc, wp_roc
 
 def create_roc_ratio(x1, y1, x2, y2):
-    sp = interpolate.interp1d(x2, y2)
+    sp = interpolate.interp1d(x2, y2, bounds_error=False, fill_value='extrapolate')
```

Comment:
In one of my cases where this was a problem, it was because one of the x-ranges was just barely outside the x-range of the other ROC curve. You probably wouldn't see the difference in the plot. But in my initial fix I added a point at 1/1 for both of the ROC curves.

Reply:
Right, extrapolation indeed can cause misinterpretation. I think the way around it would be to make sure that no curve (neither reference, nor nominal) is extrapolated beyond its domain `[x_min, x_max]` while computing the ratio. Also, I'd suggest to not interpolate the ratio curve (as it is done now), so interpolate the nominal curve instead - which kinda makes sense logically, since we want to see how nominal curve compares to working points of ratio curve (be the latter WP-like or ROC curve). Therefore, firstly one has to find a subset of points `x'` from `x2` which is also in `[min(x1), max(x1)]` (i.e. the intersection), create an interpolator for nominal curve `sp = interpolate.interp1d(x1, y1)`, apply it to `x'`: `y1_upd = sp(x')` and compute `ratio = y1_upd/y'`, where `y'` are the corresponding y-points to `x'`. Does it look reasonable?

As for the solution with adding (1, 1) point, it would still effectively be the extrapolation additionally conditioned on (1, 1) point, so I guess also not something desirable.

Reply:
Hi, I just implemented this (fix x-range and interpolate nominal curve instead reference), but in the example I tried it for, the ratio seemed wrong, because the granularity of the x-range of the reference curve was worse than in the nominal curve (See left side of https://dmroy.web.cern.ch/dmroy/plots_archive/2021_12_28_TauML/roc_curve_vs_mu_ggH_DY_pt_20_40_eta_0.0_1.5_dm_0.pdf, x-array content is [... 0.95963 0.98863 0.98955 0.99429 ...]). I was wondering, is there a reason I couldn't interpolate over both curves, after combining the x-ranges, like `x_comb = np.unique(np.sort(np.concatenate((x1, x2))))`? This looked fine on my side.

Reply:
If I understood Konstantin's concern correctly, extrapolation might be misleading because it introduces information which was never there. Of course this also applies to interpolation but to a lesser extent because ambiguity in between is bounded by neighbouring points. 

So if the ranges are combined as you proposed, what is interpolation for one curve can become extrapolation for the other. This is not a problem if both curves are ROC-like curves, because their domain covers the whole x-range [0,1]. But if the reference curve is a working point-like, then it is basically a set of points {x_1, ..., x_N} not covering [0,1]-domain and after combining with ROC-like curve x-range (i.e. [0,1]) it becomes now extrapolation for a WP-like curve.

One solution could be to make `create_roc_ratio()` aware of what is the type of the reference curve and then do interpolation depending on the type (e.g. ROC as you proposed, WP as I wrote before). But current solution looks fine to me and I am not sure if there is a need to polish things any further. @kandrosov what do you think?

Reply:
@DennRoy can you, please, share x1, y1, x2, y2 values (as they are, before any fixes) for the example where it doesn't work?

Reply:
Here they are:
https://dmroy.web.cern.ch/dmroy/plots_archive/2022_01_28_TauML/arrays

Reply:
Thank you! I can't reproduce the error. I run `create_roc_ratio` as-it-is in the master and it worked without any errors or warnings...

Reply:
perhaps it is due to precision loss when converting to string and back. Can you pickle them instead?

Reply:
There isn't any error here, but when I change the function to interpolate over x1 instead of x2 (I haven't commited this yet), then there's a loss in the accuracy for this case. You can see this in https://dmroy.web.cern.ch/dmroy/plots_archive/2022_01_28_TauML/roc_curve_vs_mu_ggH_DY_pt_20_40_eta_0.0_1.5_dm_0.pdf by directly comparing the ROC curves with the ratio between Tau efficiency of 0.98-0.99. It's because the binning is too coarse in x2. It goes from 0.95963 to 0.98863. In x1 there any many more entries within this range.